database/migrations/0001_init_podcasting_20_database.sql

```diff
@@ -0,0 +1,844 @@
+BEFORE UPDATE ON feed_flag_status
+FOR EACH ROW
+EXECUTE FUNCTION set_updated_at_field();
+
+INSERT INTO feed_flag_status (status) VALUES ('none'), ('spam'), ('takedown'), ('other'), ('always-allow');
+
+--** FEED
+
+-- The top-level table for storing feed data, and internal parsing data.
+CREATE TABLE feed (
+    id SERIAL PRIMARY KEY,
+    url varchar_url UNIQUE NOT NULL,
+
+    -- feed flag
+    feed_flag_status_id INTEGER NOT NULL REFERENCES feed_flag_status(id),
+
+    -- internal
+
+    -- Used to prevent another thread from parsing the same feed.
+    -- Set to current time at beginning of parsing, and NULL at end of parsing. 
+    -- This is to prevent multiple threads from parsing the same feed.
+    -- If is_parsing is over X minutes old, assume last parsing failed and proceed to parse.
+    is_parsing server_time,
+
+    -- 0 will only be parsed when PI API reports an update.
+    -- higher parsing_priority will be parsed more frequently on a schedule.
+    parsing_priority INTEGER DEFAULT 0 CHECK (parsing_priority BETWEEN 0 AND 5),
+
+    -- the run-time environment container id
+    container_id VARCHAR(12),
+
+    created_at server_time_with_default,
+    updated_at server_time_with_default
+);
+
+CREATE TRIGGER set_updated_at_feed
+BEFORE UPDATE ON feed
+FOR EACH ROW
+EXECUTE FUNCTION set_updated_at_field();
+
+CREATE TABLE feed_log (
+    id SERIAL PRIMARY KEY,
+    feed_id INTEGER NOT NULL REFERENCES feed(id) ON DELETE CASCADE,
+    last_http_status INTEGER,
+    last_crawl_time server_time,
+    last_good_http_status_time server_time,
+    last_parse_time server_time,
+    last_update_time server_time,
+    crawl_errors INTEGER DEFAULT 0,
+    parse_errors INTEGER DEFAULT 0
+);
+
+--** CHANNEL
+
+-- <channel>
+CREATE TABLE channel (
+    id SERIAL PRIMARY KEY,
+    id_text short_id_v2 UNIQUE NOT NULL,
+    slug varchar_slug,
+    feed_id INTEGER NOT NULL REFERENCES feed(id) ON DELETE CASCADE,
+    podcast_index_id INTEGER UNIQUE NOT NULL,
+    podcast_guid UUID UNIQUE, -- <podcast:guid>
+    title varchar_normal,
+    sortable_title varchar_short, -- all lowercase, ignores articles at beginning of title
+    medium_value_id INTEGER REFERENCES medium_value(id),
+
+    -- TODO: should we hash the last parsed feed, so we can compare it to the hash of
+    -- a feed before completely parsing it, to check if it has changed before continuing?
+
+    -- channels that have a PI value tag require special handling to request value_tag data
+    -- from the Podcast Index API.
+    has_podcast_index_value_tags BOOLEAN DEFAULT FALSE,
+
+    -- hidden items are no longer available in the rss feed, but are still in the database.
+    hidden BOOLEAN DEFAULT FALSE,
+    -- markedForDeletion items are no longer available in the rss feed, and may be able to be deleted.
+    marked_for_deletion BOOLEAN DEFAULT FALSE
+);
+
+CREATE UNIQUE INDEX channel_podcast_guid_unique ON channel(podcast_guid) WHERE podcast_guid IS NOT NULL;
```

Comment:
If you dont have a reason why `podcast_guid` would be null, just add the column with `NOT NULL` and leave `UNIQUE` on the column. This will create an index already for you.  So this would be redundant/and will fail once you run it. 

https://www.postgresql.org/docs/current/indexes-unique.html

Reply:
As mentioned in my previous comment, there are RSS feeds without the `<podcast:guid>` tag in them, so we need to save those as `null`. I was thinking that this INDEX could enforce uniqueness only in the case of not `null,` although since I have `podcast_guid UUID UNIQUE` already, I guess I should remove that `UNIQUE` since the `channel_podcast_guid_unique` will handle the proper uniqueness.